core/src/main/java/org/mule/runtime/core/internal/metadata/DefaultMapDataType.java

```diff
@@ -71,22 +73,16 @@ public boolean equals(Object o) {
       return false;
     }
 
-    // TODO MULE-9987 Fix this
-    if ((mimeType != null ? !mimeType.matches(that.mimeType) : that.mimeType != null) && !MediaType.ANY.matches(that.mimeType)
-        && !MediaType.ANY.matches(this.mimeType)) {
+    if ((this.mimeType == null && that.mimeType != null || that.mimeType == null && this.mimeType != null)) {
```

Comment:
use `Object.equals` to compare  the `mediaType`, use the getter instead of the attribute directly